File: 1046-max-consecutive-ones-iii/max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        countZeros = 0
        left = 0
        maxLength  = 0
        for right in range(len(nums)):
            countZeros += 1 - nums[right]
            while left<=right and  countZeros > k:
                  countZeros -= 1 - nums[left]
                  left+=1
            maxLength = max(maxLength, right - left + 1)
        return maxLength
            

        

File: 1046-max-consecutive-ones-iii/max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int left = 0;
        int maximumOnes = 0;
        int maximumWindowSize = 0;
        // 1,1,1,0,0,0 - if [number == 1] maximumOnes -=1
        for( int right=0; right < nums.length; right ++){
            // count maximum ones
            if ( nums[right] == 1 ) {
                maximumOnes+=1;
            }
            // window size
            int windowSize = right - left + 1;
            if ( windowSize - maximumOnes > k  ) {
                // start removing from left 
                if ( nums[left] == 1) {
                    maximumOnes-=1;
                }
                left+=1; // move pointer to exclude 0 or 1
            }
            // take the maximum elements as 1
            maximumWindowSize  = Math.max ( maximumWindowSize , right - left + 1);
        }

        return maximumWindowSize;
    }
}